features/storage/kvstore/kv_map/KVMap.cpp

```diff
@@ -0,0 +1,363 @@
+        partition->kv_config->external_fs->unmount();
+    }
+
+    if (partition->kv_config->external_bd != NULL) {
+        partition->kv_config->external_bd->deinit();
+    }
+
+    if (partition->kv_config->internal_store != NULL) {
+        partition->kv_config->internal_store->deinit();
+    }
+
+    if (partition->kv_config->internal_bd != NULL) {
+        partition->kv_config->internal_bd->deinit();
+    }
+
+    if (partition->kv_config->kvstore_main_instance != NULL) {
+        partition->kv_config->kvstore_main_instance->deinit();
+    }
+
+    delete [] partition->partition_name;
+    partition->partition_name = NULL;
+    partition->kv_config = NULL;
+}
+
+
+int KVMap::detach(const char *partition_name)
+{
+    int ret = MBED_SUCCESS;
+
+    _mutex->lock();
+
+    if (!_is_initialized) {
+        ret = MBED_ERROR_NOT_READY;
+        goto exit;
+    }
+
+    ret = MBED_ERROR_ITEM_NOT_FOUND;
+    for (int i = 0; i < _kv_num_attached_kvs; i++) {
+
+        if (strcmp(partition_name, _kv_map_table[i].partition_name) != 0) {
+            continue;
+        }
+
+        deinit_partition(&_kv_map_table[i]);
+
+        memcpy(&_kv_map_table[i], &_kv_map_table[i + 1], sizeof(kv_map_entry_t) * (MAX_ATTACHED_KVS - i - 1));
+        _kv_map_table[MAX_ATTACHED_KVS - 1].partition_name = NULL;
+        _kv_map_table[MAX_ATTACHED_KVS - 1].kv_config->kvstore_main_instance = NULL;
+        _kv_num_attached_kvs--;
+        ret = MBED_SUCCESS;
+        break;
+    }
+
+exit:
+    _mutex->unlock();
+    return ret;
+}
+
+int KVMap::deinit()
+{
+    int ret = MBED_SUCCESS;
+
+    _mutex->lock();
+
+    if (!_is_initialized) {
+        ret = MBED_ERROR_NOT_READY;
+        goto exit;
+    }
+
+    for (int i = 0; i < _kv_num_attached_kvs; i++) {
+
+        if (_kv_map_table[i].kv_config->kvstore_main_instance == NULL) {
+            goto exit;
+        }
+
+        deinit_partition(&_kv_map_table[i]);
+    }
+
+exit:
+    _kv_num_attached_kvs = 0;
+    _mutex->unlock();
+    return ret;
+}
+
+// Full name lookup and then break it into KVStore instance and key
+int KVMap::lookup(const char *full_name, KVStore **kv_instance, size_t *key_index, uint32_t *flags_mask)
+{
+    _mutex->lock();
+
+    kvstore_config_t *kv_config;
+    int ret = config_lookup(full_name, &kv_config, key_index);
+    if (ret != MBED_SUCCESS) {
+        goto exit;
+    }
+
+    *kv_instance = kv_config->kvstore_main_instance;
+    if (flags_mask != NULL) {
+        *flags_mask = kv_config->flags_mask;
+    }
+
+exit:
+    _mutex->unlock();
+    return ret;
+}
+
+// Full name lookup and then break it into KVStore configuration struct and key
+int KVMap::config_lookup(const char *full_name, kvstore_config_t **kv_config, size_t *key_index)
+{
+    int ret = MBED_SUCCESS;
+    int delimiter_index;
+    int i;
+    const char *delimiter_position;
+
+    const char *temp_str = full_name;
+
+    if (!_is_initialized) {
+        ret = MBED_ERROR_NOT_READY;
+        goto exit;
```

Comment:
Should we actually call "kv_init_storage_config" if we haven't yet? (lazy init)

Reply:
Static API is doing that. Not sure about this function because it can be called by the get_XXXX_instance that I am not sure must call kv_init_storage_config.

Reply:
I noticed get_XXXX_instance does not call kv_init_storage_config either. Should get_XXXX_instance be calling kv_init_storage_config?